### PR TITLE
Admin Implementation for navigation tree

### PR DESCRIPTION
**Pull request by webbyfox**

### CODE REVIEW

File: djangocms_navigation/forms.py [hunk 1]
```diff
@@ -1,9 +1,71 @@
...
+            choices.append(
+                (
+                    ContentType.objects.get_for_model(model).id,
+                    model._meta.verbose_name.capitalize(),
+                )
+            )
+
+            content_choices.extend([(obj.id, obj) for obj in model.objects.all()])
+
+        self.fields["content_type"].choices = choices
```
Comment (monikasulik) [inline djangocms_navigation/forms.py]: This seems quite complex for what it does really. Wondering why we're not just using `ModelChoiceField` and then doing something like `self.fields["content_type"].queryset = supported_models()` without all this looping and appending?

File: djangocms_navigation/forms.py [hunk 2]
```diff
@@ -1,9 +1,71 @@
...
+        content_choices = []
+        for model in supported_models():
+            choices.append(
+                (
+                    ContentType.objects.get_for_model(model).id,
+                    model._meta.verbose_name.capitalize(),
+                )
+            )
+
+            content_choices.extend([(obj.id, obj) for obj in model.objects.all()])
```
Comment (monikasulik) [inline djangocms_navigation/forms.py]: Won't this mean that the choice list could potentially be veeeeeeery long? And won't that potentially lead to the browser crashing (I've seen it happen)? Perhaps discuss with @chaosk ? Might make sense to have this field as an autocomplete or something like that. There are django libs that add autocomplete to the admin, so hopefully won't be too complicated, but maybe best done in a separate ticket.

File: djangocms_navigation/forms.py [hunk 3]
```diff
@@ -1,9 +1,71 @@
...
+
+        self.fields["content_type"].choices = choices
+        self.fields["object_id"].choices = content_choices
+
+        if self.instance:
+
+            if self.instance.content_type_id:
```
Comment (monikasulik) [inline djangocms_navigation/forms.py]: How come this doesn't happen automatically? I would have thought a model form would always do this?

File: djangocms_navigation/forms.py [hunk 4]
```diff
@@ -1,9 +1,71 @@
...
+                self.fields["content_type"].initial = self.instance.content_type
+                self.fields["object_id"].initial = self.instance.object_id
+
+    def clean(self):
+        data = super().clean()
+        data["content_type"] = ContentType.objects.get(id=data["content_type"])
+        return data
```
Comment (monikasulik) [inline djangocms_navigation/forms.py]: This looks quite weird. Why doesn't this happen automatically?

File: djangocms_navigation/forms.py [hunk 5]
```diff
@@ -1,9 +1,71 @@
...
+
+    def save(self, **kwargs):
+        self.instance.object_id = self.cleaned_data["object_id"]
+        self.instance.content_type = self.cleaned_data["content_type"]
```
Comment (monikasulik) [inline djangocms_navigation/forms.py]: And again?? If there's something in this class which is causing the django model form to not pick up these fields I think it's worth putting in some comments in the code as to why.

File: djangocms_navigation/models.py [hunk 6]
```diff
@@ -28,8 +28,7 @@ class Meta:
 
 class MenuContent(models.Model):
     menu = models.ForeignKey(Menu, on_delete=models.PROTECT)
-    root = models.ForeignKey(
-        'djangocms_navigation.MenuItem', on_delete=models.PROTECT)
+    root = models.ForeignKey("djangocms_navigation.MenuItem", on_delete=models.PROTECT)
```
Comment (monikasulik) [inline djangocms_navigation/models.py]: Technically speaking you've changed it from pep8 compliant to non-pep8 compliant - just sayin' ;) I know max line length isn't a big deal and there are some cases where to make the line short is to make things very difficult to read, but sometimes when doing merges in split screen etc. it's actually useful to have the lines under 80 chars.

File: djangocms_navigation/test_utils/factories.py [hunk 7]
```diff
@@ -13,10 +13,11 @@
 
 class UserFactory(factory.django.DjangoModelFactory):
     username = FuzzyText(length=12)
-    first_name = factory.Faker('first_name')
-    last_name = factory.Faker('last_name')
+    first_name = factory.Faker("first_name")
```
Comment (monikasulik) [inline djangocms_navigation/test_utils/factories.py]: hmmm, why all these changes from `'` to `"`? Both are correct, but generally most of the time when the strings are short like here people tend to use `'` and when they're sentences `"` is used more commonly.

File: djangocms_navigation/utils.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import functools
+
```
Comment (monikasulik) [inline djangocms_navigation/utils.py]: This is definitely not pep8 standard :P Both of the functools lines are standard libs so should be grouped together really

File: djangocms_navigation/utils.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+from django.apps import apps
+
+
+@lru_cache(maxsize=1)
+def supported_models():
+    app_config = apps.get_app_config("djangocms_navigation")
+    try:
+        extension = app_config.cms_extension
+        return extension.navigation_apps_models
+    except AttributeError:
```
Comment (monikasulik) [inline djangocms_navigation/utils.py]: Why is this necessary?

File: djangocms_navigation/utils.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import functools
+
+from functools import lru_cache
+
+from django.apps import apps
+
+
+@lru_cache(maxsize=1)
+def supported_models():
```
Comment (monikasulik) [inline djangocms_navigation/utils.py]: Wouldn't it make more sense if this returned a queryset of content type objects? I mean otherwise this is just an alias for `extension.navigation_apps_models`, no?

File: djangocms_navigation/utils.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    app_config = apps.get_app_config("djangocms_navigation")
+    try:
+        extension = app_config.cms_extension
+        return extension.navigation_apps_models
+    except AttributeError:
+        return app_config.navigation_apps_models
+
+
+def is_model_supported(model):
```
Comment (monikasulik) [inline djangocms_navigation/utils.py]: This doesn't appear to be used anywhere so why is it here?

File: djangocms_navigation/utils.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    """Return bool value if model is in keys"""
+    return model in supported_models().keys()
+
+
+def get_supported_model_queryset(model):
```
Comment (monikasulik) [inline djangocms_navigation/utils.py]: Ditto, not used anywhere and not sure what it does really. I don't think what it returns is actually a queryset?

File: djangocms_navigation/models.py (hunk 6 above)
Comment (webbyfox) [reply djangocms_navigation/models.py]: I use black to format code and max line length is taken from setup.cfg which is currently 120 for all our projects.

File: djangocms_navigation/admin.py [hunk 13]
```diff
@@ -1,26 +1,52 @@
...
 
 class MenuContentAdmin(admin.ModelAdmin):
-    exclude = ["menu"]
+    form = MenuContentForm
     list_display = ["title", "get_menuitem_link"]
+    list_display_links = None
 
     def save_model(self, request, obj, form, change):
         if not change:
+            title = form.data.get("title")
```
Comment (monikasulik) [inline djangocms_navigation/admin.py]: I think I missed this before... This is actually potentially a security risk - `form.data` is the data as it came in the POST. What you want to be getting is `form.cleaned_data`.

**Comment by monikasulik** [comment]
@webbyfox Are tests passing for you? Failing for me.

**Comment by webbyfox** [comment]
@monikasulik , Yes I am running `python setup.py test` and ran OK

File: djangocms_navigation/utils.py [hunk 14]
```diff
@@ -0,0 +1,13 @@
+from functools import lru_cache
+
+from django.apps import apps
+
+
+@lru_cache(maxsize=1)
+def supported_models():
+    try:
+        app_config = apps.get_app_config("djangocms_navigation")
```
Comment (monikasulik) [inline djangocms_navigation/utils.py]: It is generally bad practice to put multiple lines into a try clause unless you have a good reason for it. What can happen is that some line we weren't expecting to throw an exception does and then it gets caught and then someone spends ages debugging (have had this happen on my teams more than once). In this particular case `LookupError` is exotic enough that it's unlikely, but still good to get into good habits. AFAIK, what you probably want here is:
```
try:
    app_config = apps.get_app_config("djangocms_navigation")
except LookupError:
    return []
else:
    extension = app_config.cms_extension
    return extension.navigation_apps_models
```

File: djangocms_navigation/utils.py [hunk 15]
```diff
@@ -0,0 +1,13 @@
+from functools import lru_cache
+
+from django.apps import apps
+
+
+@lru_cache(maxsize=1)
+def supported_models():
```
Comment (monikasulik) [inline djangocms_navigation/utils.py]: Missing a test for this.

**Comment by monikasulik** [comment]
@webbyfox Although you've added validation for not allowing adding of menu items in the "root" position, it's still possible to add root menu items by adding them as siblings to existing roots (just managed to do this). So I think this is a bug.
